src/category_theory/closed/ideal.lean

```diff
@@ -0,0 +1,278 @@
+`B : D` and `A : C`, the exponential `A ⟹ B` is in `D`: resembling ring theoretic ideals. We
+define the notion here for inclusion functors `i : D ⥤ C` rather than explicit subcategories to
+preserve the principle of equivalence.
+
+We additionally show that if `C` is cartesian closed and `i : D ⥤ C` is a reflective functor, the
+following are equivalent.
+* The left adjoint to `i` preserves binary (equivalently, finite) products.
+* `i` is an exponential ideal.
+-/
+universes v₁ v₂ u₁ u₂
+
+noncomputable theory
+
+namespace category_theory
+
+open limits category
+
+section ideal
+
+variables {C : Type u₁} {D : Type u₂} [category.{v₁} C] [category.{v₁} D] {i : D ⥤ C}
+
+variables (i) [has_finite_products C] [cartesian_closed C]
+
+/--
+The subcategory `D` of `C` expressed as an inclusion functor is an *exponential ideal* if
+`B ∈ D` implies `A ⟹ B ∈ D` for all `A`.
+-/
+class exponential_ideal : Prop :=
+(exp_closed : ∀ {B}, B ∈ i.ess_image → ∀ A, (A ⟹ B) ∈ i.ess_image)
+
+/--
+To show `i` is an exponential ideal it suffices to show that `A ⟹ iB` is "in" `D` for any `A` in
+`C` and `B` in `D`.
+-/
+lemma exponential_ideal.mk' (h : ∀ (B : D) (A : C), (A ⟹ i.obj B) ∈ i.ess_image) :
+  exponential_ideal i :=
+⟨λ B hB A,
+begin
+  rcases hB with ⟨B', ⟨iB'⟩⟩,
+  exact functor.ess_image.of_iso ((exp A).map_iso iB') (h B' A),
+end⟩
+
+/-- The subcategory of subterminal objects is an exponential ideal. -/
+instance : exponential_ideal (subterminal_inclusion C) :=
+begin
+  apply exponential_ideal.mk',
+  intros B A,
+  refine ⟨⟨A ⟹ B.1, λ Z g h, _⟩, ⟨iso.refl _⟩⟩,
+  exact uncurry_injective (B.2 (cartesian_closed.uncurry g) (cartesian_closed.uncurry h))
+end
+
+/--
+If `D` is a reflective subcategory, the property of being an exponential ideal is equivalent to
+the presence of a natural isomorphism `i ⋙ exp A ⋙ left_adjoint i ⋙ i ≅ i ⋙ exp A`, that is:
+`(A ⟹ iB) ≅ i L (A ⟹ iB)`, naturally in `B`.
+The converse is given in `exponential_ideal.mk_of_iso`.
+-/
+def exponential_ideal_reflective (A : C) [reflective i] [exponential_ideal i] :
+  i ⋙ exp A ⋙ left_adjoint i ⋙ i ≅ i ⋙ exp A :=
+begin
+  symmetry,
+  apply nat_iso.of_components _ _,
+  { intro X,
+    haveI := (exponential_ideal.exp_closed (i.obj_mem_ess_image X) A).unit_is_iso,
+    apply as_iso ((adjunction.of_right_adjoint i).unit.app (A ⟹ i.obj X)) },
+  { simp }
+end
+
+/--
+Given a natural isomorphism `i ⋙ exp A ⋙ left_adjoint i ⋙ i ≅ i ⋙ exp A`, we can show `i`
+is an exponential ideal.
+-/
+lemma exponential_ideal.mk_of_iso [reflective i]
+  (h : Π (A : C), i ⋙ exp A ⋙ left_adjoint i ⋙ i ≅ i ⋙ exp A) :
+  exponential_ideal i :=
+begin
+  apply exponential_ideal.mk',
+  intros B A,
+  exact ⟨_, ⟨(h A).app B⟩⟩,
+end
+
+end ideal
+
+section
+
+variables {C : Type u₁} {D : Type u₂} [category.{v₁} C] [category.{v₁} D]
+variables (i : D ⥤ C)
+
+lemma reflective_products [has_finite_products C] [reflective i] : has_finite_products D :=
```

Comment:
Should this mover elsewhere?

Reply:
Possibly - but it's hard to say where: it has to be below `monad/limits` because that file establishes how reflective functors behave with limits, but if it were to be in that file then finite limits would need to be imported there. Do you have any suggestions?